snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/internal/tracker/Tracker.java

```diff
@@ -698,6 +701,30 @@ private void workaroundForCampaignAttributionEnrichment(@NonNull Payload payload
         }
     }
 
+    /*
+     The session context should be computed as part of `addBasicContextsToContexts` but that method
+     is executed in a separate thread. At the moment, the Session management is performed by the legacy
+     solution that doesn't make use of the tracker state snapshot (taken as soon as the event is tracked)
+     needed to keep a coherent state between events tracked in sequence but processed in parallel.
+     Due to this parallel processing the session context could be incoherent with the event sequence.
+     This limit is a serious problem when we process lifecycle events because the `application_background`
+     event should be checked with the session foreground timeout and the `application_foreground` event
+     should be checked with the session background timeout. Without a coherent session management it's
+     impossible to grant this behaviour causing serious issues on the calculation of the session expiring.
+     */
+    private void workaroundForIncoherentSessionContext(@NonNull TrackerEvent event) {
```

Comment:
This is a bit of a shame that we need to introduce the special handling, but I understand the problem. Would be great to support tracker state snapshot in the future.